syntax = "proto3";

package route;

service RoutePlannerService {
  rpc OptimizeRoute(RouteOptimizationRequest) returns (RouteOptimizationResponse);
  rpc CalculateRouteGeometry(RouteGeometryRequest) returns (RouteGeometryResponse);
}

message RouteOptimizationRequest {
  double start_lat = 1;
  double start_lon = 2;
  repeated POIInfo pois = 3;
  double available_hours = 4;
}

message POIInfo {
  int32 id = 1;
  string name = 2;
  double lat = 3;
  double lon = 4;
  int32 avg_visit_minutes = 5;
  double rating = 6;
}

message RouteOptimizationResponse {
  repeated POIInfo optimized_route = 1;
  double total_distance_km = 2;
  int32 total_minutes = 3;
}

message RouteGeometryRequest {
  double start_lat = 1;
  double start_lon = 2;
  repeated Coordinate waypoints = 3;
}

message Coordinate {
  double lat = 1;
  double lon = 2;
}

message RouteGeometryResponse {
  repeated Coordinate geometry = 1;
  double total_distance_km = 2;
}
